### PR TITLE
finagle-netty4-http: allow setting multiple values for a request cookie

### PR DESCRIPTION
**Pull request by guilgaly**
See issue #909.

Problem:
In a Response, `CookieMap` allows setting multiple cookie values for a given name as long as they differ by their domain or path. However, in a Request, if the client sent several values for a given cookie name, only the last one is currently available.

Solution:
Fix a bug in `Netty4CookieCodec` when decoding request cookies, and change `CookieMap` behavior to allow inserting multiple values for a request cookie. Also preserve insertion order for the different values for a given cookie name, to preserve the order in which the client sent the values in a request.

### CODE REVIEW

**Comment by yufangong** [comment]
@guilgaly Thanks for the PR! It looks like there are two tests failing, can you check it out? Thanks!

**Comment by guilgaly** [comment]
@yufangong it should be fixed now!

The order of iteration on the CookieMap can be different between Scala 2.12 and 2.13 because of changes in the underlying collection, and I broke these tests by no longer accepting either order - so they only passed on Scala 2.13. (My changes should guarantee that we iterate in order of insertion for cookies with the same name, but there is still no guarantee for cookies with different names.)

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (mosesn) [inline finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: suggestion:
from the RFC:
https://datatracker.ietf.org/doc/html/rfc6265#section-4.2.2
> servers SHOULD NOT rely upon the order in which these cookies appear in the header

we shouldn't make the ordering part of the contract, imo

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (mosesn) [inline finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: issue: this response vs request behavior is strange to me.  what's the motivation behind it?  I don't see anything in the RFC that would encourage it.

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (mosesn) [inline finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala:165]: question: we still want to prevent illegal duplicative cookies (eg ones with the same path, domain, and name), right?  I'm a bit worried that we might be overreaching here.  can we ensure that we continue to filter those out?

**Review by mosesn — COMMENTED** [review]
Thanks for the PR!  I think there are some details in the semantics that we should work out before merging this in.

**Comment by tigerlily-he** [comment]
Hi @guilgaly, Friendly ping. How is this PR going? Moses left some questions and comments in a review.

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (guilgaly) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: In a request, the cookie header does not provide any information on the domain/path for the cookie, only the cookie name and value; yet the cookie header can contain multiple values for the same name (presumably because they were set with different domains/paths that can both match for this request, but we don't have this information when receiving the request on the server side).

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (guilgaly) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: The way I see it:
- The RFC does also recommend that the client use a specific order in https://datatracker.ietf.org/doc/html/rfc6265#section-5.4, although that's only a "should", not a "must" (hence why it shouldn't be relied upon). Major browsers do respect that AFAIK.
- The recommendation not to rely on the order is, again only a "should not", not a "must not"; up to you to see if the HTTP library should be opinionated on that, I guess?
- I think it would be a bit surprising to have a `Seq` (rather than e.g. a `Set`) but with values in a potentially "random" order, different from the order we see when looking at the actual HTTP request.
- There may also be a question about the behavior of `get()`/`getValue()` (which use `headOption`) when several values are present. Always taking the first value in the list as sent by the client seems reasonable to me?

**Comment by guilgaly** [comment]
@tigerlily-he thanks for the ping! I must have missed the notification at the time, and after that I was mostly offline for a while and forgot to check...

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (mosesn) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: > The recommendation not to rely on the order is, again only a "should not", not a "must not"; up to you to see if the HTTP library should be opinionated on that, I guess?

I don't think we need to be opinionated about it, and for that reason, I think it's nice that the implementation preserves order, but also that we don't put into the contract that it does.  Servers can still be well-behaved and ignore the order, and it's better for people to program as if that's the case.  If people want to program in a way that's explicitly discouraged by the spec they can for now, but support for them may be dropped in the future.

> There may also be a question about the behavior of get()/getValue() (which use headOption) when several values are present. Always taking the first value in the list as sent by the client seems reasonable to me?

I agree it's reasonable, but I don't think that we should make it part of the contract.

With that said, I don't feel super strongly about this.  It's just my opinion.  I'm OK with shipping it even if you decide not to change it!

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (mosesn) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: OK, that makes sense to me, thanks!  It might be good to explain that in the scaladoc.

Comment (guilgaly) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: I'm going to try to clarify the Scaladoc then 👍

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (guilgaly) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala]: I can drop it from the stated contract in the Scaladoc, no problem with that for me.

**Comment by guilgaly** [comment]
@mosesn I modified the Scaladoc per your comments, let me know what you think. 🙂

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (hamdiallam) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala:165]: +1

**Comment by hamdiallam** [comment]
Hey! @guilgaly The PR looks good to me.

Wanted to surface Moses' question. Is there benefit to adding `message.isResponse` conditional check when adding a cookie? While it's right the on the request path there is no domain/path, is there any benefit to having multiple cookies with the same name on the request? It seems to be that it'd be beneficial to avoid duplicates on both the request and response. On the server side it's ensured that this duplication wouldn't occur, so it makes sense to simply be universal client-side as well

**Review by mosesn — APPROVED** [review]
Thanks!!

File: finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala
Comment (joybestourous) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala:165]: @guilgaly hey, just bumping this question (and @hamdiallam's related question from a few days ago)

Comment (dotordogh) [reply finagle-base-http/src/main/scala/com/twitter/finagle/http/CookieMap.scala:165]: @guilgaly bumping this once more :)

**Comment by mosesn** [comment]
@guilgaly have you had a chance to look at @hamdiallam's question?  We should be good to merge it in after you have a chance to address it!

**Comment by tigerlily-he** [comment]
Hi @guilgaly, We'd like to merge this PR after you address this question. 

> Hey! @guilgaly The PR looks good to me.
> 
> Wanted to surface Moses' question. Is there benefit to adding `message.isResponse` conditional check when adding a cookie? While it's right the on the request path there is no domain/path, is there any benefit to having multiple cookies with the same name on the request? It seems to be that it'd be beneficial to avoid duplicates on both the request and response. On the server side it's ensured that this duplication wouldn't occur, so it makes sense to simply be universal client-side as well

**Pull request closed without merging**
